MEDimage/learning/ml_utils.py

```diff
@@ -0,0 +1,844 @@
+    
+    Returns:
+        pd.DataFrame: outcome table with balanced binary labels.
+    """
+
+    # We place them prematurely in maj and min and correct it afterwards
+    n_maj = (outcome_table_binary == 0).sum().values[0]
+    n_min = (outcome_table_binary == 1).sum().values[0]
+    if n_maj == n_min:
+        return outcome_table_binary
+    elif n_min > n_maj:
+        n_min, n_maj = n_maj, n_min
+
+    # Sample the patients from the majority class
+    patient_ids_maj, patient_ids_min = get_patient_id_classes(outcome_table_binary)
+    patient_ids_min = list(patient_ids_min)
+    patient_ids_numpy = patient_ids_maj.to_numpy()
+    np.random.shuffle(patient_ids_numpy)
+    patient_ids_sample = list(patient_ids_numpy[0:n_min])
+    new_ids = patient_ids_min + patient_ids_sample
+
+    return outcome_table_binary.loc[new_ids, :]
+
+def save_model(ml: Dict, model: Dict, var_id: str, path_model: Path) -> Dict:
+    """
+    Saves a given model locally as a pickle object and outputs a dictionary
+    containing the model's information.
+
+    Args:
+        ml (Dict): The machine learning dictionary.
+        model (Dict): The model dict to save.
+        var_id (str): The stduied variable. For ex: 'var3'.
+        path_model (str): The path to save the model.
+    
+    Returns:
+        Dict: A dictionary containing the model's information.
+    """
+    # Saving model
+    with open(path_model, "wb") as f:
+        pickle.dump(model, f)
+
+    # Getting the "var_names" string
+    var_names = "__".join(ml['variables'][var_id]['nameType'])
+
+    # Recording model info
+    model_info = dict()
+    model_info['path'] = path_model
+    model_info['var_ids'] = var_id
+    model_info['var_type'] = var_names
+
+    try: # This part may fail if model training failed.
+        model_info['var_names'] = model['var_names']
+        model_info['var_info'] = model['var_info']
+        if 'normalization' in model_info['var_info'].keys():
+            if 'normalization_table' in model_info['var_info']['normalization'].keys():
+                normalization_struct = write_table_structure(model_info['var_info']['normalization']['normalization_table'])
+                model_info['var_info']['normalization']['normalization_table'] = normalization_struct
+        model_info['threshold'] = model['threshold']
+    except Exception as e:
+        print("Failed to create a fully model info")
+        print(e)
+
+    return model_info
+
+def write_table_structure(data_table: pd.DataFrame) -> Dict:
+    """
+    Writes the structure of a table in a dictionary.
+
+    Args:
+        data_table (pd.DataFrame): a table.
+    
+    Returns:
+        Dict: a dictionary containing the table's structure.
+    """
+    # Initialization
+    data_struct = dict()
+
+    if len(data_table.index) != 0:
+        data_struct['index'] = list(data_table.index)
+
+    # Creating the structure
+    for column in data_table.columns:
+        data_struct[column] = data_table[column]
```

Comment:
data_struct = {column: data_table[column] for column in data_table.columns}

if len()!=0:
    data_struct['index'] = list()